File: ios/Oriveo/Oriveo/Core/Attachments/AttachmentRouter.swift
import Foundation

nonisolated enum AttachmentRoute: Equatable, Sendable {
    case native
    case clientExtract
}

nonisolated enum AttachmentRouter {

    private static let pdfMime = "application/pdf"
    private static let scannedPdfCode = "scanned_pdf"

    static func maxNativeBytes(for provider: ProviderKind) -> Int {
        switch provider {
        case .openAI: return 30 * 1024 * 1024
        case .anthropic: return 30 * 1024 * 1024
        case .gemini: return 20 * 1024 * 1024
        default: return 30 * 1024 * 1024
        }
    }

    /// extractionErrorCode == "scanned_pdf" → native( fallback)
    ///     - model.pdfNativeDefault == true → native(Gemini)
    static func decide(
        attachment: Attachment,
        provider: ProviderKind,
        model: AIModel
    ) -> AttachmentRoute {
        guard attachment.kind == .file else { return .clientExtract }

        let mime = attachment.mimeType.lowercased()
        guard model.nativeFileMimes.contains(mime) else { return .clientExtract }

        guard let base64 = attachment.originalBase64Data, !base64.isEmpty else {
            return .clientExtract
        }

        if let bytes = attachment.extractedSizeBytes,
           bytes > 0,
           bytes > maxNativeBytes(for: provider) {
            return .clientExtract
        }

        if mime == pdfMime {
            if attachment.extractionErrorCode == scannedPdfCode {
                return .native
            }
            if model.pdfNativeDefault {
                return .native
            }
            return .clientExtract
        }

        return .native
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/attachments/AttachmentRouter.kt
package ai.oriveo.community.core.attachments

import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.Attachment
import ai.oriveo.community.core.model.AttachmentKind
import ai.oriveo.community.core.model.ProviderKind

enum class AttachmentRoute {
    Native,
    ClientExtract,
}

object AttachmentRouter {

    private const val PDF_MIME = "application/pdf"
    private const val SCANNED_PDF_CODE = "scanned_pdf"

    fun maxNativeBytes(provider: ProviderKind): Int = when (provider) {
        ProviderKind.OpenAI -> 25 * 1024 * 1024
        ProviderKind.Anthropic -> 25 * 1024 * 1024
        ProviderKind.Gemini -> 20 * 1024 * 1024
        else -> 25 * 1024 * 1024
    }

    fun decide(
        attachment: Attachment,
        provider: ProviderKind,
        model: AIModel,
    ): AttachmentRoute {
        if (attachment.kind != AttachmentKind.File) return AttachmentRoute.ClientExtract

        val mime = attachment.mimeType.lowercase()
        if (!model.nativeFileMimes.contains(mime)) return AttachmentRoute.ClientExtract

        val base64 = attachment.originalBase64Data
        if (base64.isNullOrEmpty()) return AttachmentRoute.ClientExtract

        val bytes = attachment.extractedSizeBytes ?: 0
        if (bytes > 0 && bytes > maxNativeBytes(provider)) {
            return AttachmentRoute.ClientExtract
        }

        if (mime == PDF_MIME) {
            if (attachment.extractionErrorCode == SCANNED_PDF_CODE) {
                return AttachmentRoute.Native
            }
            if (model.pdfNativeDefault) {
                return AttachmentRoute.Native
            }
            return AttachmentRoute.ClientExtract
        }

        return AttachmentRoute.Native
    }
}
